Update model-card for DINOv2 (#37104)
huggingface/transformers: docs/source/en/model_doc/dinov2.md

        <img alt="SDPA" src="https://img.shields.io/badge/SDPA-DE3412?style=flat&logo=pytorch&logoColor=white">
    </div>
</div>

# DINOv2

[DINOv2](https://huggingface.co/papers/2304.07193) is a vision foundation model that uses [ViT](./vit) as a feature extractor for multiple downstream tasks like image classification and depth estimation. It focuses on stabilizing and accelerating training through techniques like a faster memory-efficient attention, sequence packing, improved stochastic depth, Fully Sharded Data Parallel (FSDP), and model distillation.

You can find all the original DINOv2 checkpoints under the [Dinov2](https://huggingface.co/collections/facebook/dinov2-6526c98554b3d2576e071ce3) collection.

> [!TIP]
> Click on the DINOv2 models in the right sidebar for more examples of how to apply DINOv2 to different vision tasks.

The example below demonstrates how to obtain an image embedding with [`Pipeline`] or the [`AutoModel`] class.

<hfoptions id="usage">
<hfoption id="Pipeline">

```py
import torch
from transformers import pipeline

pipe = pipeline(
    task="image-feature-extraction", 

>     model="facebook/dinov2-small"

Reviewer: ```suggestion
    model="facebook/dinov2-small",
    torch_dtype=torch.float16,
    device=0
```

)

embeddings = pipe("http://images.cocodataset.org/val2017/000000039769.jpg")
image_embedding = torch.tensor(embeddings[0]).mean(dim=0)
```

</hfoption>
<hfoption id="AutoModel">

```py
import requests
import torch
from transformers import AutoImageProcessor, AutoModel
from PIL import Image

url = 'http://images.cocodataset.org/val2017/000000039769.jpg'
image = Image.open(requests.get(url, stream=True).raw)

processor = AutoImageProcessor.from_pretrained('facebook/dinov2-base')

> model = AutoModel.from_pretrained('facebook/dinov2-base')

Reviewer: ```suggestion
model = AutoModel.from_pretrained("facebook/dinov2-base", torch_dtype=torch.float16, device_map="auto", attn_implementation="sdpa")
```

inputs = processor(images=image, return_tensors="pt")
image_embedding = model(**inputs)[0][0].mean(dim=0)
```

> </hfoption>

Reviewer: ```suggestion
</hfoption>
</hfoptions>
```

## Notes

> - [`DINOv2 with Registers`] 

Reviewer: ```suggestion
```

- Use [torch.jit.trace](https://pytorch.org/docs/stable/generated/torch.jit.trace.html) to speedup inference. However, it will produce some mismatched elements. The difference between the original and traced model is 1e-4.

> ```py

Reviewer: Indent this code block so its aligned under the list item here

        <img alt="FlashAttention" src="https://img.shields.io/badge/%E2%9A%A1%EF%B8%8E%20FlashAttention-eae0c8?style=flat">
        <img alt="SDPA" src="https://img.shields.io/badge/SDPA-DE3412?style=flat&logo=pytorch&logoColor=white">
    </div>
</div>

# DINOv2

> [DINOv2](https://huggingface.co/papers/2304.07193) is a vision foundation model that is used as a feature extractor for multiple downstream tasks like image classification and depth estimation. The training consists of a student and a teacher model that have the same [`Vision Transformer (ViT)`] architecture. The student model is optimized with gradient-descent to produce outputs similar to the teacher model. Embeddings of uncurated images are matched with embeddings of curated images through k-NN, resulting in a larger dataset for pretraining.

Reviewer: ```suggestion
[DINOv2](https://huggingface.co/papers/2304.07193) is a vision foundation model that uses [ViT](./vit) as a feature extractor for multiple downstream tasks like image classification and depth estimation. It focuses on stabilizing and accelerating training through techniques like a faster memory-efficient attention, sequence packing, improved stochastic depth, Fully Sharded Data Parallel (FSDP), and model distillation.
```

> You can find all the original DINOv2 checkpoints under the [Dinov2](https://huggingface.co/collections/facebook/dinov2-6526c98554b3d2576e071ce3) release.

Reviewer: ```suggestion
You can find all the original DINOv2 checkpoints under the [Dinov2](https://huggingface.co/collections/facebook/dinov2-6526c98554b3d2576e071ce3) collection.
```

> [!TIP]
> Click on the DINOv2 models in the right sidebar for more examples of how to apply DINOv2 to different vision tasks.

The example below demonstrates how to obtain an image embedding with [`Pipeline`] or the [`AutoModel`] class.

<hfoptions id="usage">
<hfoption id="Pipeline">

```py

> import torch

Reviewer: I think for both the `Pipeline` and `AutoModel` examples, it'd be more practical to show an image classification example.

We can also show a quantization example with the facebook/dinov2-giant checkpoint!

Reviewer: Yeah we can do something like this with a finetuned checkpoint:

```py
import torch
from transformers import pipeline

pipe = pipeline(
    task="image-classification", 
    model="facebook/dinov2-small-imagenet1k-1-layer",
    torch_dtype=torch.float16,
    device=0
)

pipe("https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/pipeline-cat-chonk.jpeg")
```

Reviewer: Thanks, I think you may have forgotten a the quantization example with the https://huggingface.co/facebook/dinov2-giant-imagenet1k-1-layer checkpoint though!

Reviewer: This still needs to be addressed ☝️

A list of official Hugging Face and community (indicated by 🌎) resources to help you get started with DINOv2.
    processor = AutoImageProcessor.from_pretrained('facebook/dinov2-base')
    model = AutoModel.from_pretrained('facebook/dinov2-base')

- Demo notebooks for DINOv2 can be found [here](https://github.com/NielsRogge/Transformers-Tutorials/tree/master/DINOv2). 🌎
    inputs = processor(images=image, return_tensors="pt")
    outputs = model(**inputs)
    last_hidden_states = outputs[0]

<PipelineTag pipeline="image-classification"/>
    # We have to force return_dict=False for tracing
    model.config.return_dict = False

- [`Dinov2ForImageClassification`] is supported by this [example script](https://github.com/huggingface/transformers/tree/main/examples/pytorch/image-classification) and [notebook](https://colab.research.google.com/github/huggingface/notebooks/blob/main/examples/image_classification.ipynb).
- See also: [Image classification task guide](../tasks/image_classification)
    with torch.no_grad():
        traced_model = torch.jit.trace(model, [inputs.pixel_values])
        traced_outputs = traced_model(inputs.pixel_values)

> If you're interested in submitting a resource to be included here, please feel free to open a Pull Request and we'll review it! The resource should ideally demonstrate something new instead of duplicating an existing resource.

Reviewer 2: Providing code snippets is great, however it would be super nice to keep resources within docs, they usually provide more in-depth view of the model, e.g. finetuning 🤗 

cc @stevhliu

model = AutoModel.from_pretrained('facebook/dinov2-base')

inputs = processor(images=image, return_tensors="pt")
image_embedding = model(**inputs)[0][0].mean(dim=0)
```

</hfoption>

## Notes

> - [`DINOv2 with Registers`] 

Reviewer: Add indented code snippet of tracing under here.

```suggestion
- Use [torch.jit.trace](https://pytorch.org/docs/stable/generated/torch.jit.trace.html) to speedup inference. However, it will produce some mismatched elements. The difference between the original and traced model is 1e-4.

   <code>
```
